### ISSUE TITLE
Problem with FRefine

### ISSUE THREAD

**Issue by jbschroder** [XBraid/xbraid#21]
Using the OpenMPI that I have with Ubuntu fails some regression tests, which didn't fail on my tux-box at LLNL (or on my macbook right now).

Essentially, ex-01-refinement seems to work with 1 proc, but sometimes fail with multiple procs.

A simple way to generate this issue is to change line 118 in ex-01-refinement.c to
   if ((level == 0) && (nrefine < 2))
And then to run 
  mpirun -np 1 ex-01-refinement -nt 100 -tol 1e-6 -refine 2 -max_rfactor 2 -mi 5
  mpirun -np 2 ex-01-refinement -nt 100 -tol 1e-6 -refine 2 -max_rfactor 2 -mi 5
  mpirun -np 5 ex-01-refinement -nt 100 -tol 1e-6 -refine 2 -max_rfactor 2 -mi 5

The problem gets worse with larger numbers of MPI tasks, which is a bit un-nerving.  

With the above setup, if you switch to print_level=3, you can see the first discrepancy show up at with np=2, and mi=2, at point 198 (at the interface between proc 0 and proc 1).  When things break, for np=2, the residual at point 198 is 0.  For np=1, it has a "normal" value compared to surrounding points.

For the case of np=5, this problem of a "zero" residual then repeats at regular intervals of about every 80 time points (at processor boundaries).

**Comment by jbschroder** [comment]
I suppose this could be an issue with my MPI installation, but it's just the standard OpenMPI from the package manager...

**Comment by jbschroder** [comment]
So, I tracked down this little memcheck bug on Quartz, and got that fixed.  But, I still get weird (spurious) results on my UNM tuxbox for ex-01-refinement, as outlined above

**Comment by rfalgout** [comment]
Any idea which commit caused this?  Was it recent or has this been lurking around for a while now?

**Comment by jbschroder** [comment]
I don't know which commit caused this, and I only see it on my UNM tuxbox, not my Mac, not on Quartz, and not on my old LLNL tuxbox.  The memory bug that I found on Quartz did not resolve things on my UNM tuxbox.

**Comment by jbschroder** [comment]
Hmmm...I could give you an account on my UNM tuxbox, and you could try replicating it

**Comment by jbschroder** [comment]
OK.  I have an update here.  I did finally figure out that this was an issue with my MPI implementation.  Here's the digest for the record.

The problem was sending and receiving f_next in FRefine.  For this case that fails (see above post for runstring), processor 1 is sending and receiving f_next to itself.  The Irecv is initiated

         MPI_Irecv(&f_next, 1, braid_MPI_INT, MPI_ANY_SOURCE, 3, comm, &request);

and then the send is done

            MPI_Send(&msg, 1, braid_MPI_INT, prevproc, 3, comm);

and finally, the Wait is done later at 

   if (f_npoints > 0)
   {
      if (f_iupper < f_gupper)
      {
           MPI_Wait(&request, &status);
      }
   }

By adding fprintf(stderr, "...")  statements, I was able to see processor 1 initiate the correct send for f_next to itself, but then verify that the MPI_Wait() completes, without the data having arrived.  Weird and sad.

I did some reading, and pairing MPI_Irecv() and MPI_Send() should be OK for sending a message to yourself.  So, I'm not thinking that a coding change needs to happen in Braid.  I've only seen this issue with this one particular old OpenMPI implementation.  So, I'll close this ticket here in a minute.

Epilogue:  Well..that just ate 2 days of my life to track down -- arghh....  Perhaps a morning beer is in order :-)
